File: programmers.co.kr/hash/2.cpp
// https://programmers.co.kr/learn/courses/30/lessons/42577
// Phonebook

#include <string>
#include <unordered_map>
#include <vector>

using namespace std;

bool solution(vector<string> phone_book) {
    unordered_map<string, int> map;
    for (string phoneNum : phone_book) {
        map[phoneNum]++;
    }
    for (string phoneNum : phone_book) {
        for (size_t i = 1; i < phoneNum.length(); i++) {
            string sub = phoneNum.substr(0, i);
            if (map.find(sub) != map.end()) {
                return false;
            }
        }
    }
    return true;
}

File: programmers.co.kr/hash/2.py
# https://programmers.co.kr/learn/courses/30/lessons/42577?language=python3

import itertools


def solution(phone_book):
    for i, __ in enumerate(phone_book):
        for j, num in enumerate(phone_book):
            if i != j and phone_book[i].startswith(phone_book[j]):
                return False

    return True
